From 1e87a7f80dbe1a2fb3b3fe9569fcddef0ed43753 Mon Sep 17 00:00:00 2001
From: Zhonghu Xu <xuzhonghu@huawei.com>
Date: Wed, 18 May 2022 16:27:41 +0800
Subject: Only build workloadInstance when se selects pod enabled (#38925)

* Disable se selects pods

* Optimization: skip building workloadinstance when no workload handler registered

* Add release note

* Revert PILOT_ENABLE_SERVICEENTRY_SELECT_PODS

* Revert PILOT_ENABLE_SERVICEENTRY_SELECT_PODS
---
 pilot/pkg/serviceregistry/kube/controller/multicluster.go | 1 -
 pilot/pkg/serviceregistry/kube/controller/pod.go          | 4 ++++
 2 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/pilot/pkg/serviceregistry/kube/controller/multicluster.go b/pilot/pkg/serviceregistry/kube/controller/multicluster.go
index 17b83fe868..af8d2c9f35 100644
--- a/pilot/pkg/serviceregistry/kube/controller/multicluster.go
+++ b/pilot/pkg/serviceregistry/kube/controller/multicluster.go
@@ -169,7 +169,6 @@ func (m *Multicluster) ClusterAdded(cluster *multicluster.Cluster, clusterStopCh
 
 	m.m.Unlock()
 
-	// TODO move instance cache out of registries
 	if m.serviceEntryController != nil && features.EnableServiceEntrySelectPods {
 		// Add an instance handler in the kubernetes registry to notify service entry store about pod events
 		kubeRegistry.AppendWorkloadHandler(m.serviceEntryController.WorkloadInstanceHandler)
diff --git a/pilot/pkg/serviceregistry/kube/controller/pod.go b/pilot/pkg/serviceregistry/kube/controller/pod.go
index 4a9e494860..07030f5d45 100644
--- a/pilot/pkg/serviceregistry/kube/controller/pod.go
+++ b/pilot/pkg/serviceregistry/kube/controller/pod.go
@@ -167,6 +167,10 @@ func (pc *PodCache) onEvent(curr interface{}, ev model.Event) error {
 
 // notifyWorkloadHandlers fire workloadInstance handlers for pod
 func (pc *PodCache) notifyWorkloadHandlers(pod *v1.Pod, ev model.Event) {
+	// if no workload handler registered, skip building WorkloadInstance
+	if len(pc.c.handlers.GetWorkloadHandlers()) == 0 {
+		return
+	}
 	// fire instance handles for workload
 	ep := NewEndpointBuilder(pc.c, pod).buildIstioEndpoint(pod.Status.PodIP, 0, "", model.AlwaysDiscoverable)
 	workloadInstance := &model.WorkloadInstance{
-- 
2.35.3

